### PR TITLE
Updated Azure Storage libraries to v12

### PR DESCRIPTION
**Pull request by valeneiko**
## Description
Updated Azure Storage library to v12, because v11 is deprecated

```
Microsoft.Azure.Storage.Blob (v11.2.2) ==> Azure.Storage.Blobs (v12.8.0)
```

## Related issues
Addresses [issue #].

## Testing
Existing tests pass.

## FHIR Team Checklist
- [ ] **Update the title** of the PR to be succinct and less than 50 characters
- [ ] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [ ] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [ ] Tag the PR with **Azure API for FHIR** if this will release to the managed service
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationClient/AzureAccessTokenClientInitializer.cs [hunk 1]
```diff
@@ -7,42 +7,37 @@
...
-        private readonly IAccessTokenProvider _accessTokenProvider;
+        private readonly TokenCredential _credential;
         private readonly ExportJobConfiguration _exportJobConfiguration;
-        private readonly ILogger<AzureAccessTokenClientInitializer> _logger;
 
         public AzureAccessTokenClientInitializer(
-            IAccessTokenProvider accessTokenProvider,
-            IOptions<ExportJobConfiguration> exportJobConfiguration,
-            ILogger<AzureAccessTokenClientInitializer> logger)
+            TokenCredential credential,
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Azure/ExportDestinationClient/AzureAccessTokenClientInitializer.cs:25]: Since we do not register a service of type `TokenCredential` the framework will fail to resolve it and hence cannot create an `AzureAccessTokenClientInitializer` object.

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationClient/AzureAccessTokenClientInitializer.cs [hunk 2]
```diff
@@ -7,42 +7,37 @@
...
+            IOptions<ExportJobConfiguration> exportJobConfiguration)
         {
-            EnsureArg.IsNotNull(accessTokenProvider, nameof(accessTokenProvider));
+            EnsureArg.IsNotNull(credential, nameof(credential));
             EnsureArg.IsNotNull(exportJobConfiguration?.Value, nameof(exportJobConfiguration));
-            EnsureArg.IsNotNull(logger, nameof(logger));
 
-            _accessTokenProvider = accessTokenProvider;
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Azure/ExportDestinationClient/AzureAccessTokenClientInitializer.cs:35]: Is there a specific reason we do not want to use this interface anymore. The idea behind it was to allow an easy way to change the logic of acquiring access tokens by injecting a different implementation of `IAccessTokenProvider`.

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/GetExportRequestHandler.cs [hunk 3]
```diff
@@ -58,7 +58,7 @@ public async Task<GetExportResponse> Handle(GetExportRequest request, Cancellati
                 var jobResult = new ExportJobResult(
                     outcome.JobRecord.QueuedTime,
                     outcome.JobRecord.RequestUri,
-                    requiresAccessToken: false,
+                    requiresAccessToken: true,
```
Comment (anyname2) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/Export/GetExportRequestHandler.cs:61]: This should not be part of the PR

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs [hunk 4]
```diff
@@ -635,7 +635,7 @@ private async Task ProcessSearchResultsAsync(IEnumerable<SearchResultEntry> sear
         {
             // Update the continuation token in local cache and queryParams.
             // We will add or udpate the continuation token in the query parameters list.
-            progress.UpdateContinuationToken(Convert.ToBase64String(System.Text.Encoding.UTF8.GetBytes(continuationToken)));
+            progress.UpdateContinuationToken(continuationToken);
```
Comment (anyname2) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs:638]: This should not be part of the PR

File: src/Microsoft.Health.Fhir.Core.UnitTests/Features/Operations/Export/ExportJobTaskTests.cs [hunk 5]
```diff
@@ -1259,7 +1259,7 @@ public async Task GivenAGroupExportJobToResume_WhenExecuted_ThenAllPatientResour
                     {
                         // The ids aren't in the query parameters because of the reset
                         ids = new string[] { "1", "2", "3" };
-                        continuationTokenIndex = int.Parse(Encoding.UTF8.GetString(Convert.FromBase64String(x.ArgAt<IReadOnlyList<Tuple<string, string>>>(1)[2].Item2)).Substring(2));
+                        continuationTokenIndex = int.Parse(x.ArgAt<IReadOnlyList<Tuple<string, string>>>(1)[2].Item2.Substring(2));
```
Comment (anyname2) [inline src/Microsoft.Health.Fhir.Core.UnitTests/Features/Operations/Export/ExportJobTaskTests.cs:1262]: This should not be part of the PR

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationClient/AzureAccessTokenClientInitializer.cs (hunk 1 above)
Comment (anyname2) [reply src/Microsoft.Health.Fhir.Azure/ExportDestinationClient/AzureAccessTokenClientInitializer.cs:25]: Good catch! I forgot to include that change in the PR.

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationClient/AzureAccessTokenClientInitializer.cs (hunk 2 above)
Comment (anyname2) [reply src/Microsoft.Health.Fhir.Azure/ExportDestinationClient/AzureAccessTokenClientInitializer.cs:35]: AFAIK, all the azure libraries are now using `Azure.Core` for auth, hence `TokenCredential` is accepted as a parameter. I didn't see an easy way to use `IAccessTokenProvider` here.
Also `TokenCredential` is an abstract class, so in the sense of injecting different implementation, you can still do that.

Comment (anyname2) [reply src/Microsoft.Health.Fhir.Azure/ExportDestinationClient/AzureAccessTokenClientInitializer.cs:35]: Actually, there is even a diclaimer that it should not be used anymore: https://docs.microsoft.com/en-us/dotnet/api/overview/azure/service-to-service-authentication

**Comment by namadabu** [comment]
@anyname2 Apologies for the late response. Thanks for bringing this up. This SDK update is something we definitely need to do. But this update also introduces breaking changes and we have to update our build/test pipelines and our managed service to handle this change. We are planning on working on this later rather than now. @CaitlinV39 FYI

**Comment by rbans96** [comment]
Closing this out as currently we don't have any plan in near future to update azure storage libraries but we can definitely refer this PR whenever we plan this in long run.
